metadata-ingestion/src/datahub/ingestion/source/state/stale_entity_removal_handler.py

```diff
@@ -217,11 +218,30 @@ def gen_removed_entity_workunits(self) -> Iterable[MetadataWorkUnit]:
         ):
             # Log the failure. This would prevent the current state from getting committed.
             self.source.get_report().report_failure(
-                "Stateful Ingestion",
+                "stale-entity-removal",
                 f"Will not soft-delete entities, since we'd be deleting {entity_difference_percent:.1f}% of the existing entities. "
                 f"To force a deletion, increase the value of 'stateful_ingestion.fail_safe_threshold' (currently {self.stateful_ingestion_config.fail_safe_threshold})",
             )
-            # Bail so that we don't emit the stale entity removal workunits.
+            return
+
+        if self.source.get_report().events_produced == 0:
+            # SUBTLE: By reporting this as a failure here, we also ensure that the
+            # new (empty) state doesn't get committed.
+            # TODO: Move back to using fail_safe_threshold once we're confident that we've squashed all the bugs.
+            self.source.get_report().report_failure(
+                "stale-entity-removal",
+                "Skipping stale entity soft-deletion because the source produced no events. "
+                "This is a fail-safe mechanism to prevent accidental deletion of all entities.",
+            )
+            return
+
+        # If the source already had a failure, skip soft-deletion.
+        # TODO: Eventually, switch this to check if anything in the pipeline had a failure so far.
+        if self.source.get_report().failures:
```

Comment:
Does this take into account the `commit_policy`? Or for stale entity removal the rule is just commit iff there are no failures

Reply:
Does that mean we're dropping support for a configurable commit policy? Right now it seems like it doesn't really do anything

Reply:
eventually yes I'd like to drop support for that